File: hw/palomo/u3/hw26Observer/Observer/src/main/java/ec/edu/espe/observer/model/Stock.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */

package ec.edu.espe.observer.model;

import java.util.ArrayList;
import java.util.Iterator;
/**
 *
 * @author Cristian Palomo, Error 404, @ESPE
 */
public abstract class Stock {
    protected String symbol;
    protected double price;
    private ArrayList<Investor> investors = new ArrayList<>();

    public Stock() {}

    public void addObserver(Investor investor) {
        investors.add(investor);
    }

    public void deleteObserver(Investor investor) {
        investors.remove(investor);
    }

    public void notifyObservers(Object args) {
        Iterator<Investor> i = investors.iterator();
        while (i.hasNext()) {
            Investor investor = i.next();
            investor.update(this, args);
        }
    }

    public String getSymbol() {
        return symbol;
    }

    public void setSymbol(String symbol) {
        this.symbol = symbol;
        notifyObservers(symbol);
    }

    public double getPrice() {
        return price;
    }

    public void setPrice(double price) {
        this.price = price;
        notifyObservers(Double.valueOf(price));
    }
}

File: hw/palomo/u3/hw26Observer/Observer.py/ec_edu_espe/observer/model/stock.py
"""
@author Cristian Palomo, Error 404, @ESPE
"""
from abc import ABC
from typing import List, Optional

from ec_edu_espe.observer.model.investor import Investor


class Stock(ABC):

    def __init__(self) -> None:
        self.symbol: Optional[str] = None
        self.price: Optional[float] = None
        self._investors: List[Investor] = []

    def add_observer(self, investor: Investor) -> None:
        self._investors.append(investor)

    def delete_observer(self, investor: Investor) -> None:
        if investor in self._investors:
            self._investors.remove(investor)

    def notify_observers(self, args: object) -> None:
        for investor in list(self._investors):
            investor.update(self, args)

    def get_symbol(self) -> Optional[str]:
        return self.symbol

    def set_symbol(self, symbol: str) -> None:
        self.symbol = symbol
        self.notify_observers(symbol)

    def get_price(self) -> Optional[float]:
        return self.price

    def set_price(self, price: float) -> None:
        self.price = price
        self.notify_observers(float(price))
